File: java/behavioral/InterpreterDemo.java
//author sbkun
package behavioral;

import java.util.Vector;

class Context {
    Context(String input) {
        this.input = input;
    }

    public String getInput() {
        return input;
    }

    public void setInput(String input) {
        this.input = input;
    }

    public String getOutput() {
        return output;
    }

    public void setOutput(String output) {
        this.output = output;
    }

    private String input;
    private String output;
}

abstract class Expression {
    abstract void interpret(Context context);
}

class IntToHexExpression extends Expression{
    @Override
    void interpret(Context context) {
        String arg = context.getInput();
        if (arg.contains("Hexadecimal")) {
            int intArg = Integer.parseInt(arg.substring(0, arg.indexOf(" ")));
            context.setOutput(Integer.toHexString(intArg));
        }
    }
}

class IntToBinaryExpression extends Expression{

    @Override
    void interpret(Context context) {
        String arg = context.getInput();
        if (arg.contains("Binary")) {
            int intArg = Integer.parseInt(arg.substring(0, arg.indexOf(" ")));
            context.setOutput(Integer.toBinaryString(intArg));
        }
    }
}

class InterpreterClient {
    InterpreterClient() {
        //build the parse tree
        interpreters.add(new IntToBinaryExpression());
        interpreters.add(new IntToHexExpression());
    }

    String interpret(Context context) {
        for(Expression e: interpreters) {
            e.interpret(context);
        }
        return context.getOutput();
    }

    private Vector<Expression> interpreters = new Vector<>();
}

public class InterpreterDemo {
    public static void main(String args[]) {
        String str1 = "28 in Binary";
        String str2 = "28 in Hexadecimal";

        InterpreterClient interpreterClient = new InterpreterClient();

        Context context1 = new Context(str1);
        Context context2 = new Context(str2);

        String ret1 = interpreterClient.interpret(context1);
        String ret2 = interpreterClient.interpret(context2);

        System.out.println(str1+" = "+ret1);
        System.out.println(str2+" = "+ret2);
    }
}

File: cpp/behavioral/InterpreterDemo.cpp
//author sbkun
#include "noncopyable.h"

#include <iostream>
#include <string>
#include <iomanip>
#include <vector>
#include <sstream>

using namespace std;

string intToBinaryString(unsigned long n);

class Context {
public:
	explicit Context(const string& input) : input_(input) {
	}
	
	const string& getInput() const {
		return input_;
	}
	
	void setOutput(const string& output) {
		output_ = output;
	}	
	
	string getOutput() const {
		return output_;
	}	
private:
	string input_;	
	string output_;
};

class Expression: noncopyable {
public:
	virtual ~Expression() = default;
	virtual void interpret(Context& context) = 0;
};

class Int2Hex: public Expression{
public:
	void interpret(Context& context) override {
		const string& input= context.getInput();
		if (input.find("Hexadecimal") != string::npos) {
				int intArg = stoi(input.substr(0, input.find(" ")));	
				std::stringstream stream;
				stream << std::hex << intArg;
				context.setOutput(stream.str());
		}
	}
};

class Int2Binary: public Expression{
public:
	void interpret(Context& context) override {
		const string& input= context.getInput();
		if (input.find("Binary") != string::npos) {
				int intArg = stoi(input.substr(0, input.find(" ")));	
				context.setOutput(intToBinaryString(intArg));
		}
	}
};

class InterprtClient {
public:
	InterprtClient(){
		expressions_.emplace_back(new Int2Binary);
		expressions_.emplace_back(new Int2Hex);
	}

	void interpret(Context& context) {
		for(auto& e: expressions_) {
			e->interpret(context);
		}
	}

private:
	vector<unique_ptr<Expression>> expressions_;
};

string intToBinaryString(unsigned long n)
{
	char     bArray[ (sizeof( unsigned long ) * 8) + 1 ];
	unsigned index  = sizeof( unsigned long ) * 8;
	char temp =0;
	bArray[ index ] = '\0';
	do{          
		temp = (n & 1); 
		temp = temp + '0';
		bArray[ --index ] = temp;
	}while (n >>= 1);
	return string( bArray + index );
}

int main(int argc, char* argv[])
{
	string str1 = "28 in Binary";
	string str2 = "28 in Hexadecimal";

	Context context1(str1);
	Context context2(str2);

	InterprtClient client;
	client.interpret(context1);
	client.interpret(context2);

	cout<<str1<<" = "<<context1.getOutput()<<endl<<endl;
	cout<<str2<<" = "<<context2.getOutput()<<endl<<endl;

	return 0;
}
